File: solution/day11/src/main/java/roman/numerals/RomanNumerals.java
package roman.numerals;

import java.util.Comparator;
import java.util.Map;
import java.util.Optional;
import java.util.TreeMap;

import static java.util.Optional.empty;
import static java.util.Optional.of;

public final class RomanNumerals {
    private static final int MAX_NUMBER = 3999;
    private static final Map<Integer, String> intToNumerals = createMapForIntegerToNumerals();

    private RomanNumerals() {
    }

    private static TreeMap<Integer, String> createMapForIntegerToNumerals() {
        var map = new TreeMap<Integer, String>(Comparator.reverseOrder());
        map.put(1000, "M");
        map.put(900, "CM");
        map.put(500, "D");
        map.put(400, "CD");
        map.put(100, "C");
        map.put(90, "XC");
        map.put(50, "L");
        map.put(40, "XL");
        map.put(10, "X");
        map.put(9, "IX");
        map.put(5, "V");
        map.put(4, "IV");
        map.put(1, "I");

        return map;
    }

    public static Optional<String> convert(int number) {
        return isInRange(number)
                ? convertSafely(number)
                : empty();
    }

    private static Optional<String> convertSafely(int number) {
        var roman = new StringBuilder();
        var remaining = number;

        for (var toRoman : intToNumerals.entrySet()) {
            while (remaining >= toRoman.getKey()) {
                roman.append(toRoman.getValue());
                remaining -= toRoman.getKey();
            }
        }
        return of(roman.toString());
    }

    private static boolean isInRange(int number) {
        return number > 0 && number <= MAX_NUMBER;
    }
}

File: exercise/c#/Day11/Day11/RomanNumerals.cs
using System.Text;
using LanguageExt;
using static LanguageExt.Option<string>;

namespace Day11
{
    public static class RomanNumerals
    {
        private const int MaxNumber = 3999;

        private static readonly IReadOnlyDictionary<int, string> IntToNumerals = new Dictionary<int, string>()
        {
            {1000, "M"},
            {900, "CM"},
            {500, "D"},
            {400, "CD"},
            {100, "C"},
            {90, "XC"},
            {50, "L"},
            {40, "XL"},
            {10, "X"},
            {9, "IX"},
            {5, "V"},
            {4, "IV"},
            {1, "I"},
        };

        public static Option<string> ToRoman(this int number) => Convert(number);

        private static Option<string> Convert(int number)
            => IsInRange(number)
                ? ConvertSafely(number)
                : None;

        private static string ConvertSafely(int number)
        {
            var roman = new StringBuilder();
            var remaining = number;

            foreach (var toRoman in IntToNumerals)
            {
                while (remaining >= toRoman.Key)
                {
                    roman.Append(toRoman.Value);
                    remaining -= toRoman.Key;
                }
            }

            return roman.ToString();
        }

        private static bool IsInRange(int number) => number is > 0 and <= MaxNumber;
    }
}